Clean Const Function Pointer Removal

thk123 opened the pull request:
A fairly naive approach to optimising function pointers in some specific common cases. If a function pointer is:

- constant and assigned from one of:
- directly from a function symbol
- assigned to a constant FP
- a component of a const struct
- an element in a constant array *
- A constant pointer pointing to something
- Or some combination of the above

We replace it with either a direct function call, or generate the list of actually possible functions. This is opposed to generating a list of all functions with valid signatures. 

There was a rough version of this in #434, this takes the same idea, splits it out in to a separate class. It also deals with more cases and should be more robust to chaining of weird things. I have also renamed the large body of tests from fp-removal1,...fp-removal45 to meaningful names describing what they test. 

Issues: diffblue/cbmc-toyota#12 [private]
This also includes the improvments outlined in #476

thk123 commented:
There are a bunch of commits relating to the old test names, I could just remove them, but there may be some useful notes in the commit messages, so I think I will just bulk squash them even though the files they relate to don't exist in the final commit. Particularly in light of the fact git has picked it up as a rename. 


Other things I am aware of before this PR is ready for review:
- [ ] I also will go through the tests and make sure they are consistent with each other. 
- [ ] Splitting up the two big functions in remove_const_function_pointers.cpp (as done with the ID_index case)

There is also a reasonable amount of duplication between `resolve_expression` and `resolve_function_call`. In fact, the main difference is what it does with the results. I imagine there is a way of extracting the common stuff in to the `resolve_expression` and have the `resolve_function_call` use this.

thk123 commented:
This currently isn't building due to a compile error in musketeer, investigating.

thk123 commented:
@martin-cs raised an issue (#524) about dealing with null pointers. There is a flag called "--pointer-check" which will add an assertion after the IFs in case it didn't match any. To validate this works correctly with my tests I should:

 - [ ] Add the flag to all the tests
 - [ ] The tests with null having the correct handling
 - [ ] Adding some tests for the different ways the FP can be assigned null

tautschnig commented on src/goto-programs/Makefile:
```diff
@@ -19,6 +19,7 @@ SRC = goto_convert.cpp goto_convert_function_call.cpp \
       class_hierarchy.cpp show_goto_functions.cpp get_goto_model.cpp \
       slice_global_inits.cpp goto_inline_class.cpp class_identifier.cpp \
       show_goto_functions_json.cpp \
+      remove_const_function_pointers.cpp \
```
Cf the discussion on coding guidelines: your insertion point is a random one (which is fine at present), but for the principle of least surprise there should be a canonical insertion point.

tautschnig commented on src/goto-programs/remove_const_function_pointers.h:
```diff
@@ -0,0 +1,96 @@
+class remove_const_function_pointerst:public messaget
+{
+public:
+  typedef std::set<exprt> functionst;
```
Is there a reason for it not to be a std::unordered_set?

thk123 replied:
I didn't think there was but having changed it, I can think of a reason, though I don't know if it qualifies as good: the tests have to become a lot less precise as the ordering of the functions is no longer deterministic.

tautschnig replied:
I'd suggestion one of two options here: 1) the tests really shouldn't be depending on ordering, thus try to adjust the tests; 2) or add a comment saying that using `std::set` simplifies testing.

thk123 replied:
Well I had thought it was required to make the tests actually test anything, but actually checking for the existence of the goto in the no-match and approx cases will test to the same precision so I will do that.

thk123 replied:
Ah no I realise where this will cause problems - in the approx tests as they will pass even if no match. They can however be fixed without ordering by adding the cases that shouldn't be generated to the exclude lines.

tautschnig commented on src/goto-programs/remove_const_function_pointers.h:
```diff
@@ -0,0 +1,96 @@
+  bool try_resolve_typecast_function_call(
+    const typecast_exprt &typecast_expr, functionst &out_functions);
+
+  // recursive functions for dealing with the auxillary elements
```
Typo: "auxillary" -> auxiliary

tautschnig commented on src/goto-programs/remove_const_function_pointers.h:
```diff
@@ -0,0 +1,96 @@
+    bool &out_is_const);
+
+  bool is_expression_const(const exprt &expression);
+  bool is_type_const(const typet &type);
```
Likely those two functions should be marked const? Also, `is_const_expression`, `is_const_type` would seem more consistent (e.g., replace_const_symbol).

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -14,12 +14,14 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "remove_skip.h"
 #include "remove_function_pointers.h"
 #include "compute_called_functions.h"
+#include <goto-programs/remove_const_function_pointers.h>
```
Use `#include "remove_const_function_pointers.h"` for consistency with the preceding ones.

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -362,40 +367,55 @@ void remove_function_pointerst::remove_function_pointer(
+  bool found_functions=fpr(functions);
+
+  // Consistency checks
+  if(found_functions && functions.size()==0)
```
`functions.empty()` (.size() is linear time, even in C++11 I believe)

thk123 replied:
For [`std::set`](http://en.cppreference.com/w/cpp/container/set/size)/[`std::unordered_set`](http://en.cppreference.com/w/cpp/container/unordered_set/size) (which is was `functions` is) `size()` is constant time. None the less, I think `empty()` expresses the check clearer so I will do this.

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -362,40 +367,55 @@ void remove_function_pointerst::remove_function_pointer(
+  }
 
-  bool return_value_used=code.lhs().is_not_nil();
+  if(functions.size()==1)
```
Hmm, ok, if you need the actual size: maybe just compute it once.

thk123 commented:
@tautschnig Addressed your feedback (including the unordered set)
@martin-cs I've adjusted and added tests for dealing with null pointers through the use of `--pointer-check` but the code hasn't changed (well apart from [adding support in goto-analyze for `--pointer-check`](https://github.com/diffblue/cbmc/pull/519/commits/23819a28b01f73fb5a33583d68927a6a4ff7b734)), so for this to fix your problem you'll need to run goto-analyze with --pointer-check.

thk123 commented:
Before merging, there is an issue diffblue/cbmc-toyota#27 to add a flag to activate this FP removal mode only, which is probably small enough to just be part of this PR.

thk123 commented:
Note: before merging fix the approx-* tests: https://github.com/diffblue/cbmc/pull/519#discussion_r100823365

tautschnig commented:
I'm going to take a look at the entire PR, but would then wonder whether some squashing should be applied.

thk123 commented:
I've tried to squash as much as I can. However, all the further squashing I could think of became very fiddly and time consuming to apply. As such I abandoned it as I didn't think it was worth the time it would take. For example, renaming the `is_expr_const` -> `is_const_expr`, new calls get added in multiple commits so each would have to be edited.

tautschnig commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1012 @@
+Function: remove_const_function_pointerst::try_resolve_index_of_function_call
+
+  Inputs:
+   index_expr - The index expression to  to resolve to possible function calls
```
"to  to"

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -46,6 +50,14 @@ class remove_function_pointerst
+  // subset of possible functions and just leave the function pointer
+  // as it is.
+  // This can be activated in goto-instrument using --remove-const-fps
+  // instead of remove--function-pointers
```
That's the "--" in the wrong place.

thk123 commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1012 @@
+        symbol_table.lookup(expression.get(ID_identifier));
+      if(symbol.type.id()!=ID_code)
+      {
+        const exprt &symbol_value=symbol.value;
```
So the answer here is we don't. No attempt is made to validate that const-ness has been respected. 

I initially believed that we didn't need to worry at all about UB and that if the code was doing something illegal, it was acceptable to not worry about it. However, since discussion about null pointers (#524), I am less certain. 

Ultimately, I don't have enough knowledge to make a judgement on this. I was under the impression that we can't easily validate the const-ness before first resolving the FPs resulting in a chicken/egg problem. Unfortunately @martin-cs is away this week, perhaps @danpoe knows more?

thk123 replied:
@tautschnig @danpoe has suggested as a simple fix for getting this done without introducing a gap we scan the GOTO program for any casts of FPs that remove const-ness and if any are found, we abandon removing FPs in this way. That way for the majority of code that doesn't cast away const-ness we can still make the improvement without having todo anything too arduous but we are still correct if there is such a cast and it matters.

tautschnig replied:
That certainly is a good option. There might be gaps, but it would yield a defined point for inserting additional sanity checks.

thk123 commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1012 @@
+
+  Inputs:
+   expr - The expression to get the possible function calls
+   out_functions - The functions this expression could be
```
I don't think so - the expression is what is being called rather than doing the calling. I've replaced it with 
> the functions this expression could be resolved to

Is that clearer?

thk123 commented:
@tautschnig Apart from the question re: modifying const variables I believe I have addressed all of your feedback. We may have to wait till next week to resolve the const problem unfortunately. 

Apologies for all the typos and spelling mistakes, I am investigating how I can turn on a spell checker in my IDE since these are just as annoying for reviewers as linting errors!

tautschnig commented:
@thk123 Thanks a lot for the incredibly quick turnarounds on all change requests! I do agree that all questions are addressed, with the exception of the const-ness check. It would be great to have that included.

thk123 commented:
@tautschnig I have added a check for when the code contains a cast away from const and some tests to verify it. It could probably do with a more experienced eye in case I've missed a case as I wasn't sure about the way of checking the expression tree. Commit: https://github.com/diffblue/cbmc/pull/519/commits/423b34dc6cf7fc6a3164813e5f9626221a8adbbb

thk123 commented:
Tests are failing again for some reason, I'm sure they passed locally. It looks to be something to do with the test.pl change - the output looks fine.

forejtv commented:
https://github.com/diffblue/cbmc/blob/150e353ab00bca829404a38de8a5f107e7decbaa/regression/goto-instrument/approx-array-variable-const-fp-only-remove-const/test.desc has unescaped `(` and `)`. I guess the other failing tests will have similar problems.

thk123 commented:
Rebased to fix the linting issues.
@forejtv Fixed the build issues (they were in a different folder to the one I had fixed) so should pass this time, just waiting on that and a review from @tautschnig on the new changes (https://github.com/diffblue/cbmc/pull/519/commits/423b34dc6cf7fc6a3164813e5f9626221a8adbbb)

peterschrammel commented on regression/goto-analyzer/approx-array-variable-const-fp/main.c:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  return 0;
+}
```
missing newline at end of file (and other files)

thk123 replied:
This should now be fixed

peterschrammel commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -889,6 +889,7 @@ bool cbmc_parse_optionst::process_goto_program(
     // remove function pointers
     status() << "Removal of function pointers and virtual functions" << eom;
     remove_function_pointers(
+      *message_handler,
```
use `get_message_handler()` (several instances)

peterschrammel commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1013 @@
+#include <util/simplify_expr.h>
+#include <util/arith_tools.h>
+
+
```
remove one blank line

peterschrammel commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1013 @@
+
+  if(resolved)
+  {
+    out_functions.insert(resolved_functions.begin(), resolved_functions.end());
```
Is `try_resolve_function_call` expected to be called with non-empty `out_functions` arguments?

thk123 replied:
No - should I assert this at the start of the function?

thk123 replied:
I added an assert for this.

peterschrammel commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1013 @@
+
+  if(!resolved)
+  {
+    LOG("Could not resolve arary", index_expr);
```
array?

peterschrammel commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1013 @@
+          is const, or the values of the array are const.
+
+ Purpose: To squash an index access by first finding the array it is accessing
+          Then if the index can be resovled, return the squashed value. If
```
resolved

peterschrammel commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1013 @@
+            {
+              for(const exprt &resolved_array_entry : array_contents)
+              {
+                if(resolved_array_entry .is_zero())
```
too many braces for my taste

thk123 replied:
Removing braces introduce bugs like this:https://github.com/diffblue/cbmc-toyota/pull/63#pullrequestreview-27401804 so I'd really rather not do this...

peterschrammel commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1013 @@
+          The out_is_const will return whether the object that gets
+          dereferenced is constant.
+
+ Purpose: To squash an dereference access by first finding the address_of
```
a dereference

tautschnig commented on regression/goto-analyzer/no-match-const-fp-const-cast/main.c:
```diff
@@ -0,0 +1,35 @@
+{
+  const void_fp fp = f2;
+
+  // Warning: this looses const-ness of f2
```
Nit: looses -> loses

tautschnig commented on regression/goto-analyzer/no-match-const-fp-const-cast/main.c:
```diff
@@ -0,0 +1,35 @@
+  func();
+
+  return 0;
+}
```
Missing newline

thk123 replied:
Fixed

tautschnig commented on regression/goto-analyzer/no-match-const-fp-const-cast2/main.c:
```diff
@@ -0,0 +1,35 @@
+{
+  const void_fp fp = f2;
+
+  // Warning: this looses const-ness of f2
```
Nit: looses -> loses

tautschnig commented on regression/goto-analyzer/no-match-const-fp-const-cast2/main.c:
```diff
@@ -0,0 +1,35 @@
+  func();
+
+  return 0;
+}
```
Missing newline

thk123 replied:
Fixed

tautschnig commented on regression/goto-analyzer/no-match-const-pointer-const-struct-const-fp-const-cast/main.c:
```diff
@@ -0,0 +1,46 @@
+  container_pointer_modifier->fp = f4;
+
+  // Illegal:
+  //meta_fp = &f4;
```
What would `meta_fp` be?

thk123 replied:
Not sure, I think this must be a copy paste job, will remove.

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -376,30 +382,41 @@ void remove_function_pointerst::remove_function_pointer(
+  else
+  {
+    remove_const_function_pointerst fpr(
     *message_handler, pointer, ns, symbol_table);
```
I'm not sure whether github is just hiding whitespace changes at this moment, else the indent needs to be fixed.

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -376,30 +382,41 @@ void remove_function_pointerst::remove_function_pointer(
+    {
+      warning() << "Reported optimized function pointer call, but didn't"
+                << " find any functions" << eom;
+    }
```
Isn't this an internal error (which should this be an assertion)?

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -376,30 +382,41 @@ void remove_function_pointerst::remove_function_pointer(
+    {
+      warning() << "Reported didn't optimize function pointer call, but did"
+                << " find some functions to replace with" << eom;
+    }
```
So what should happen in this case? Is there something the user can do?

thk123 replied:
Nope you're right, these should both be assertions.

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -521,6 +538,110 @@ void remove_function_pointerst::remove_function_pointer(
+        return true;
+      }
+
+      // see if the rhs looses const inside the expression tree
```
nit: looses -> loses

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -521,6 +538,110 @@ void remove_function_pointerst::remove_function_pointer(
+        // Types equality does not check, for example, const-ness
+        // If this is true, then this expression only modified the
+        // type by some qualifier (or not at all)
+        if(pre_op_type==post_op_type)
```
On types, use `base_type_eq` instead of plain equality

thk123 replied:
OK, warning, I have seen this elsewhere: https://github.com/diffblue/cbmc/blob/master/src/ansi-c/c_typecheck_expr.cpp#L187 don't know if that should be fixed or is different.

tautschnig replied:
True, but note that the code over there does filter out `type1.id()==ID_symbol` (and `type2.id()==ID_symbol`) early on. So what you need to be careful about in comparing types is that some part of them may require lookups in the symbol table. `base_type_eq` takes care of this for you.

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -521,6 +538,110 @@ void remove_function_pointerst::remove_function_pointer(
+bool remove_function_pointerst::is_type_at_least_as_const_as(
+  typet type_more_const, typet type_compare)
+{
+  while(type_compare.id()!="" && type_more_const.id()!="")
```
Use `empty()` instead of string equality

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -521,6 +538,110 @@ void remove_function_pointerst::remove_function_pointer(
+            return true;
+          }
+        }
+        rhs=rhs.op0();
```
Is it worth noting that both type casts and dereferencing have only a single operand? Are there any other expressions that may need traversal (I don't think so).

thk123 replied:
Could you elaborate on this comment? This assertion fails : `assert(rhs.id()==ID_typecast || rhs.id()==ID_dereference);` because we, for example, might assign from an array of constant function pointers to a non-constant pointer. Or are you suggesting these can be ignored since only dereference and typecast can lose const (e.g. halt at the first `exprt` that isn't an typecast or dereference.

tautschnig replied:
My apologies for having been unclear. What I'm concerned about is that, without any further check, `op0()` is invoked and it's assumed that following this branch of the expression tree is the right thing to do. What I am looking for is a justification for this unconditional operation (or, else, a test to limit it to those expressions that permit following this approach).

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -521,6 +538,110 @@ void remove_function_pointerst::remove_function_pointer(
 
 /*******************************************************************\
 
+Function: remove_function_pointerst::does_program_contain_const_removal_cast
```
This doesn't seem to be specific to function pointers. I would thus opt to have it as a separate analysis, just like dirtyt or is_threaded.

thk123 replied:
Done.

thk123 commented:
Some of the tests are called `derefence` but should be `dereference`

thk123 commented:
@danpoe  Rebased for release 11 (feedback still outstanding for merge into master).

martin-cs commented:
@tautschnig could you have a look at this and see if more work is needed.

thk123 commented:
@tautschnig @martin-cs Sorry there is still outstanding changes from Michael's last review, I am hoping to get to them this week.

thk123 commented on regression/goto-instrument/approx-array-variable-const-fp-remove-all-fp/main.c:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  return 0;
+}
```
Missing new line

thk123 replied:
Fixed

thk123 commented on regression/goto-instrument/no-match-non-const-fp-only-remove-const/main.c:
```diff
@@ -0,0 +1,31 @@
+  func();
+
+  return 0;
+}
```
Missing new lines

thk123 replied:
Fixed

thk123 commented on regression/goto-instrument/no-match-non-const-fp-remove-all-fp/main.c:
```diff
@@ -0,0 +1,31 @@
+  func();
+
+  return 0;
+}
```
Missing new line

thk123 replied:
Fixed

thk123 commented on regression/goto-instrument/precise-const-fp-only-remove-const/main.c:
```diff
@@ -0,0 +1,30 @@
+  func();
+
+  return 0;
+}
```
Missing new line

thk123 replied:
Fixed

thk123 commented on regression/goto-instrument/precise-const-fp-remove-all-fp/main.c:
```diff
@@ -0,0 +1,30 @@
+  func();
+
+  return 0;
+}
```
Missing new line

thk123 replied:
Fixed.

tautschnig commented:
@thk123 Please just let me know when you're done with all edits as I'll then do another review.

thk123 commented on src/analyses/does_remove_const.cpp:
```diff
@@ -0,0 +1,138 @@
+            return true;
+          }
+        }
+        rhs=rhs.op0();
```
Moving on going discussion here:
> tautschnig:
> Is it worth noting that both type casts and dereferencing have only a single operand? Are there any other expressions that may need traversal (I don't think so).

> thk123:
> Could you elaborate on this comment? This assertion fails : assert(rhs.id()==ID_typecast || rhs.id()==ID_dereference); because we, for example, might assign from an array of constant function pointers to a non-constant pointer. Or are you suggesting these can be ignored since only dereference and typecast can lose const (e.g. halt at the first exprt that isn't an typecast or dereference.

> tautschnig
> My apologies for having been unclear. What I'm concerned about is that, without any further check, op0() is invoked and it's assumed that following this branch of the expression tree is the right thing to do. What I am looking for is a justification for this unconditional operation (or, else, a test to limit it to those expressions that permit following this approach).

thk123 replied:
OK so I think I've done this by modifying the while loop to only descend `ID_typecast` and this seems to work (I added some more tests and they all seem fine), but I am having trouble convincing myself theoretically. For this not to work we need some C code that produces an AST where the root is not a typecast and the two sides have the same const-ness, but somewhere in the RHS the const-ness is lost. 

```
void_fp x = //(something that must also be non const)
                               \
                             probably a typecast 
                                              \ 
                                         something const
```

Maybe this isn't possible as I can't think how a typecast can appear anywhere except at the root and it's immediate descendent (or in some sub-expression like an index but that isn't related. Can you think of anything? Going to explore what happens with pointer arithmetic to see if anything can be constructed in that way.

tautschnig replied:
I think the `?:` operator would be an example (the `,` operator another one, but that one will have been rewritten). For `?:` the first operand is of an type unrelated to that of the other operands.

thk123 replied:
Yeah so this assignment breaks by only descending `ID_typecast`:

```C
  // Warning: this loses const-ness of f2
  void_fp * p2fp = ((void_fp*)&fp) + 0;
```

And indeed if I swap it around, the original code doesn't even work since it is actually `op1` we need to descend. 

I think we need to check all operands with the same fundamental type as the root, does that feel like a robust approach?

tautschnig commented on src/analyses/does_remove_const.cpp:
```diff
@@ -0,0 +1,138 @@
+    {
+      const code_assignt assign=to_code_assign(instruction.code);
+      typet rhs_type=assign.rhs().type();
+      typet lhs_type=assign.lhs().type();
```
While I'm at it: all of those should be (const) references.

thk123 replied:
Done.

thk123 commented:
@tautschnig I have addressed your feedback, I think this is ready for final review. Particular attention should be payed to this commit: https://github.com/diffblue/cbmc/pull/519/commits/dea60db2e7336470492fad61c10cee20e1e0069c that deals with our discussion about descending the expression tree in the does_remove_const analysis.

tautschnig commented on src/analyses/does_remove_const.cpp:
```diff
@@ -0,0 +1,166 @@
+  {
+    if(instruction.is_assign())
+    {
+      const code_assignt assign=to_code_assign(instruction.code);
```
That should probably be a const reference

tautschnig commented on src/analyses/does_remove_const.cpp:
```diff
@@ -0,0 +1,166 @@
+      {
+        return true;
+      }
+    }
```
I'm always a bit scared when I see too many closing parentheses, so here is what *I* would have done. This is *not* a request for changes, but just my personal style: 1) Use `if(!instruction.is_assign()) continue;` above. 2) Drop the braces around `return true;` 3) Not use a temporary to store the return value.

Only make changes if you share any of those feelings.

thk123 replied:
I have made some of the changes, but can't bring myself to remove braces...

tautschnig commented on src/analyses/does_remove_const.cpp:
```diff
@@ -0,0 +1,166 @@
+\*******************************************************************/
+
+bool does_remove_constt::is_type_at_least_as_const_as(
+  typet type_more_const, typet type_compare) const
```
To avoid lots of temporary objects being created I'd actually use pointers here.

thk123 replied:
_This might be a load of rubbish but_ I thought `irept` were copy on write so therefore creating copies was cheap?

tautschnig replied:
An irept's data member is copy-on-write. That still implies allocating the encompassing object. Having looked at too many profiles, each individual such call surely is cheap, but 10^10 of them aren't anymore. Note that such a copy also induces increasing the reference count, and the destruction induces the corresponding decrement. Each of those operations is unnecessary.

tautschnig commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1013 @@
+bool remove_const_function_pointerst::try_resolve_expression(
+  const exprt &expr, expressionst &out_resolved_expression, bool &out_is_const)
+{
+  const exprt &simplified_expr=simplify_expr(expr, ns);
```
`simplify_expr` returns by value, thus you don't really want the reference here.

tautschnig commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1013 @@
+  if(resolved)
+  {
+    if(index_value_expressions.size()==1 &&
+      index_value_expressions.front().id()==ID_constant)
```
Indent by one more space

thk123 replied:
You mean the line should be indented 3 more than the line above? I thought the guidelines said always two more than the previous line?

tautschnig commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1013 @@
+            {
+              for(const exprt &resolved_array_entry : array_contents)
+              {
+                if(resolved_array_entry .is_zero())
```
There seems to be a spurious space before `.is_zero`

tautschnig commented on src/goto-programs/remove_const_function_pointers.cpp:
```diff
@@ -0,0 +1,1013 @@
+              {
+                if(resolved_array_entry .is_zero())
+                {
+                  continue;
```
Can't this be simplified quite a bit by just negating the condition on the `if`?

tautschnig commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -338,41 +356,74 @@ void remove_function_pointerst::remove_function_pointer(
-  bool return_value_used=code.lhs().is_not_nil();
+    // Reported didn't optimize function pointer call, but did find some
+    // functions to replace with
+    assert(found_functions || functions.empty());
```
Aren't those two assertions saying `assert(found_functions!=functions.empty());`

thk123 replied:
I knew there was a way to simplify this, I wanted to use a (non-existent) implication logical operator but you're right it is an if and only if relation so this works.

thk123 commented:
Regarding the pointer change, I got tripped up by the fact that that const and non-const version of `subtype` return different result based on whether the value is const. I don't know if this is good, I would expect the non-const version to create a copy of the nil data structure and return that.

thk123 commented:
@zemanlx @forejtv any idea what it means for this part of the travis build to be failing:

https://travis-ci.org/diffblue/cbmc/builds/213778812

It is the docker run build(?) and it appears to stall when it hits running the regression tests.

zemanlx commented:
`goto-cc` delete whole system in recursion during running tests. @NathanJPhillips is working on fix https://github.com/diffblue/cbmc/pull/672

NathanJPhillips commented:
That fix is merged. Rebase on master to get it.

thk123 commented:
@zemanlx I've rebased after @NathanJPhillips said but the build is still failing but there is nothing in the log, any ideas?

NathanJPhillips commented:
The log is empty, that's a different problem. Pushing a commit to restart it might make it go away.

thk123 commented:
@tautschnig Fixed whitespace as requested - this I believe is good to go.

tautschnig commented:
Assigned to @kroening for final review and merging.
